libraries/AP_BattMonitor/AP_BattMonitor_MAV.cpp

```diff
@@ -0,0 +1,108 @@
+		_state.healthy = false;
+		return;
+	} 
+	_state.healthy = true;	
+   return;
+}
+
+bool AP_BattMonitor_MAVLink_Battery::capacity_remaining_pct(uint8_t &cap_remain_pct) const
+{
+    if (_capacity_rem_pct == INT8_MIN) {
+		return false;
+	}
+	/* positive capacity remaining is expected */
+	if (_capacity_rem_pct > 0) {
+		cap_remain_pct = _capacity_rem_pct;
+	} else {
+		cap_remain_pct = 0;
+	}
+	return true;
+}
+
+bool AP_BattMonitor_MAVLink_Battery::get_cycle_count(uint16_t &cycles) const
+{
+    if (_cycle_count != UINT16_MAX) {
+		cycles = _cycle_count;
+    	return true;
+	}
+	return false;
+}
+
+void AP_BattMonitor_MAVLink_Battery::process_info_msg(const mavlink_message_t &msg)
+{
+	int32_t capacity = mavlink_msg_smart_battery_info_get_capacity_full(&msg);
+	if (capacity != -1) {
+		_params._pack_capacity.set_and_notify(capacity);
+	}
+	_cycle_count = mavlink_msg_smart_battery_info_get_cycle_count(&msg);
+	const uint16_t min_discharge_mV = mavlink_msg_smart_battery_info_get_discharge_minimum_voltage(&msg);
+	if (min_discharge_mV != UINT16_MAX) {
+		_params._critical_voltage.set_default(min_discharge_mV); 
+	}
+}
+
+void AP_BattMonitor_MAVLink_Battery::process_status_msg(const mavlink_message_t &msg)
+{
+	__mavlink_battery_status_t bat_stat;
+	mavlink_msg_battery_status_decode(&msg, &bat_stat);
+	_fault_bitmask = bat_stat.fault_bitmask;
+	uint32_t volt_sum_mV = 0;
+	const uint8_t cell_count_mav = MAVLINK_MSG_BATTERY_STATUS_FIELD_VOLTAGES_LEN;
+    const uint8_t cell_count_mav_ext = MAVLINK_MSG_BATTERY_STATUS_FIELD_VOLTAGES_EXT_LEN;
+    const uint8_t cell_count = MIN(cell_count_mav + cell_count_mav_ext, AP_BATT_MONITOR_CELLS_MAX);
+    for (uint8_t i= 0; i<cell_count; i++) {
+	    const uint16_t voltage = (i<cell_count_mav) ? bat_stat.voltages[i] : bat_stat.voltages_ext[i-cell_count_mav];
+	    if (voltage != UINT16_MAX) {
+		    volt_sum_mV += voltage ;
+	    }
+	    _state.cell_voltages.cells[i] = voltage;
+    }
+	_state.voltage = (float) volt_sum_mV * 0.001f;
+	_state.current_amps = (float) bat_stat.current_battery * 0.01f;
+	_state.consumed_mah = (float) bat_stat.current_consumed;
+    _state.consumed_wh = (float) bat_stat.energy_consumed * 0.027777778f;
+	if (bat_stat.temperature == INT16_MAX) {
+		_have_temp = false;
+	} else {
+    	_state.temperature = (float) bat_stat.temperature * 0.01f;
+		_have_temp = true;
+	}
+	if (bat_stat.time_remaining < 0) {
+		bat_stat.time_remaining *= -1;									/* if negative: time until charged */
+	}
+	_state.time_remaining = (uint32_t) bat_stat.time_remaining;
+	_state.has_time_remaining = (bool) bat_stat.time_remaining;
+	_state.last_time_micros = AP_HAL::micros();
+	_capacity_rem_pct = bat_stat.battery_remaining;
+}
+
+void AP_BattMonitor_MAVLink_Battery::handle_mavlink_battery_message(const mavlink_message_t &msg) {
+	if (msg.sysid != mavlink_system.sysid || (msg.compid != MAV_COMP_ID_BATTERY && msg.compid != MAV_COMP_ID_BATTERY2)) {
+		return;
+	}

```

Comment:
And another check to return if serial number param is>=0 and doesn't match the battery ID.